### PR TITLE
Integration of pytest

### PR DESCRIPTION
**Pull request by polybassa**
This is just a PoC. A *.uts file will be transformed to a pytest like file. The exported file usually needs manual modifications, but this exporter can be a first step towards pytest

### CODE REVIEW

**Comment by gpotter2** [comment]
Please don't merge this as is.
This is a great script but I'd like to see how well it works (a fully converted file) before doing anything.

I'm still unsure whether it's best to move to pytest because of how massive the work will probably end up being..

**Comment by polybassa** [comment]
This is absolutely not ready for merge. Sorry for not mentioning this. 

The produced files will all need manual rework. 

This is just a starting point. The use of pytest needs to be evaluated for the tests, in my opinion. 
-- 
Viele Grüße

Nils Weiß

File: scapy/tools/UTscapy.py [hunk 1]
```diff
@@ -963,21 +989,21 @@ def main():
         if six.PY2:
             KW_KO.append("python3_only")
             if VERB > 2:
-                print("### Python 2 mode ###")
+                print("### Python 2 mode ###", file=sys.stderr)
```
Comment (akorb) [inline scapy/tools/UTscapy.py]: Bei Python 2 gibt es das `file` Argument noch nicht:

```
>>> print('test', file=sys.stderr)
  File "<stdin>", line 1
    print('test', file=sys.stderr)
                      ^
SyntaxError: invalid syntax
```

**Comment by polybassa** [comment]
I've added pytest to tox (just for PoC, not added everywhere). What about supporting pytest and UTscapy in parallel?

**Comment by polybassa** [comment]
@gpotter2 @guedou @p-l- 
I would like to start a discussion. I've made a small PoC to enable pytest for the unit tests.
I respect the individual configurations by parsing for the argument "-K"
A simple dump function in UTscapy allows to transfer a *.uts file to a pytest file. Manual fixing is required here, nevertheless this script should save some time if one wants to port tests.
I've tested codecov if the combination of tests work corretly, if pytest and uts files for the same unit differ. 

Would it make sense to open the unit test system to allow contributors to write tests in *.uts or pytest?

**Comment by guedou** [comment]
This is a really cool PR. It its definitely worth finishing it and merging. If @gpotter2 and @p-l- are OK, I think that we should move to pytest and leave UTScapy behind us. After merging this PR, the next step will be to mandate that new unit tests are written with pytest while we migrate all tests.

Did you try concurrent testing? I will be interested to know if moving to pytest will also make the testing time shorter.

**Comment by polybassa** [comment]
Thanks for your feedback. I will continue my work on this PR. I will probably start to port regression.uts to pytest, to see if this concept holds. 
I didn't tested concurrent execution, yet. Also my tests on CAN will probably fail if they were executed concurrently. But I will test this, after I ported some tests from regression.uts.

**Comment by gpotter2** [comment]
I agree with @guedou that this is very cool, and it would be great to be able to migrate to pytest, but we must make sure everything works before doing that. Thé `UTscapy` file is filled with tons of extra code that would need to be integrated somehow to our test process.

**Comment by polybassa** [comment]
I totally agree, the UTscapy file is pretty much filled with stuff. I did a rough analysis of the current functionality.. in my opinion, some function might not be useful, some functions will be available in pytest. Most important is the tag system, which I could "somehow" port to pytest. 
But I will continue on that PoC, than we can see, where leads to.

**Comment by guedou** [comment]
Before converting moving to pytest, I think that it will make sense to split `regression.uts` into smaller files that respect `scapy/` naming convention i.e. `test/config.uts` `test/layers/inet6.uts` `test/packet.uts` ...

First, it will make your life easier when adding or looking for a unit test, then we will be able to convert to `pytest` file by file with several simple PR instead of a huge one.

**Comment by p-l-** [comment]
That is really great. Thanks a lot for this work.

I agree with @guedou that we may want so split `regression.uts`.

File: scapy/tools/UTscapy.py (hunk 1 above)
Comment (polybassa) [reply scapy/tools/UTscapy.py:1004]: ```from __future__ import print_function``` does this.

**Comment by polybassa** [comment]
Hi, I've updated this PR and would like to open it for your review. 

I made some first ports to pytest. I used the Co-Author tag to preserve the original authors (```git blame filename --porcelain | grep "^author " | sort | uniq``` was used to identify the original authors).

I would suggest the enforcement of flake8 for unit tests. The possibility of pytest would help us to have much cleaner unit tests, in my opinion.

**Comment by polybassa** [comment]
And another question. Should there be a copyright notice on unit tests?

**Comment by polybassa** [comment]
I will collect all ports of `*.uts` files to `test_*.py` files here in individual commits

File: test/scapy/layers/test_dhcp6.py [hunk 2]
```diff
@@ -0,0 +1,1940 @@
+import pytest
+from scapy.all import *
+
+
+# DUID_LLT basic instantiation
+def test_DUIDLLTbasicinstantiation():
```
Comment (gpotter2) [inline test/scapy/layers/test_dhcp6.py:6]: Don't you think it would be cleaner to have the comment as a docstring rather than a comment?

**Comment by polybassa** [comment]
Absolutely. Thanks for this idea.
-

**Comment by polybassa** [comment]
I've added your suggestion

**Comment by polybassa** [comment]
Does it make sense to continue porting individual unit test, as I did already in this PR? I'm a little bit afraid, that changes to uts files that are already ported in this PR get lost.

**Comment by gpotter2** [comment]
The best thing would be that you wait for all the files to be split before wasting time fixing those files. We need to have this done file by file otherwise the history will be a mess.

**Comment by gpotter2** [comment]
Just to be clear, this is **incredible work**. it's just a bit too early

**Comment by polybassa** [comment]
Ok, perfect. Thanks.

I'll try to assist with splitting regression.uts

**Comment by guedou** [comment]
I made a quick and dirty script that automates the conversion and maintains the the unit tests authorship in the commits. The commits for scapy/layers are ready but I am not pushing them to avoid conflicts.

I will try to open an issue soon in order to synchronize.

**Comment by polybassa** [comment]
Alright. Than I just wait :-)

**Comment by guedou** [comment]
@polybassa I don't know what is the best way to sync on this topic. Could you do a first PR to split VXLAN tests out of `regression.uts` using suggestions from https://github.com/secdev/scapy/pull/2755#pullrequestreview-464408077 and the PR that I recently made?

**Comment by polybassa** [comment]
Of course. I suppose the following:
1. I make a minimal PR just to add the configurations for pytest
2. I make one PR for the migration of VXLAN utscapy to pytest

Does this align with your intention?

**Comment by guedou** [comment]
I meant moving the vxlan tests out of regression.uts as I recently did for other protocols. I still consider that it is too early to use pytest and that we need to split regression.uts first.

An alternative to pytest could be to the cw the uts in parallel

**Comment by polybassa** [comment]
Ah alright..  I misunderstood. I will look into vxlan tests and try to split them.
-- 
Viele Grüße

Nils Weiß

**Comment by polybassa** [comment]
@guedou See #2913

**Comment by polybassa** [comment]
@guedou should I continue with extracting layers and "modules" from regression.uts? I saw 
- Scapy version
- Packetlist
- MPLS
- Random
- Fields
- ARP
- inet
 and some more.

**Comment by guedou** [comment]
We have an issue with Travis right now: we ran out free credits to launch the unit tests. I think that we need to focus on mandatory fixes first =/

I am taking care of inet.py and l2.py.

What is the issue with MPLS? It seems that that test are located in the correct location.

**Comment by polybassa** [comment]
Ok, so I will not do further splits of regression, until I get a sign from you. 
Regarding MPLS, I found this section in regression.uts: https://github.com/secdev/scapy/blob/c09354817d268c31674549082b9305ece6b1ad66/test/regression.uts#L4851-L4916

**Comment by guedou** [comment]
That's definitely something that we should move =)
Good catch!

**Comment by polybassa** [comment]
@guedou What are the next steps to get forward with pytest support?

**Comment by guedou** [comment]
I was very excited to be able to switch to pytest to make the tests run faster, but to be honest I'm less and less convinced that using pytest is a good idea:
- converting unit tests to python script is not straightforward. It will be highlty time consuming and will imply many careful reviews
- speed improvements won't be significative
- UTScapy and pytest should not coexist

I am working on some UTScapy improvements that will help on the performance side, and will likely make part of the test suite execute in paralell.

**Comment by polybassa** [comment]
Thanks for your answer. I can understand all your points. In terms of performance you are probably right. The main advantages I see in pytest compared to UTScapy are:

- Tests are python scripts, which means, 
    - blank lines and indentations are supported
    - typing can be applied 
    - flake8 can be used on tests
    - codespell can be used on tests
- Most IDEs have support of pytest, which allows for extensive refactoring features

In my opinion, these additions to the test system will lead us to cleaner and more maintainable tests in the future. 

Furthermore, pytest comes with a whole bunch of extensions.

A main problem, that you also mentioned, is the transition from UTScapy to pytest. If there is no way for you to have both systems coexisting for a certain period, for example one release cycle, I totally agree, that a huge effort is required to make a fast switch. Which is almost not affordable for a open source project. 

To summarize my opinions, if coexistence of two test systems for a certain time is a no go, we can stop here with pytest.
If not, I would be happy if we could make a plan for a smooth transition over a fixed period, for example the next release cycle.

**Comment by polybassa** [comment]
closed, because its broken

**Comment by gpotter2** [comment]
Honnestly I think it's a decent idea, and maybe one day we'll be happy to have it as a basis.
It's just too much work for now, but who knows 😛

**Comment by gpotter2** [comment]
Honnestly I think it's a decent idea, and maybe one day we'll be happy to have it as a basis.
It's just too much work for now, but who knows 😛

**Comment by evverx** [comment]
> The main advantages I see in pytest compared to UTScapy

I'd add that pytest should make it much easier to catch resource warnings like https://github.com/secdev/scapy/issues/3962#issuecomment-1495793735 as well as other warnings produced by `python3 -Xdev -Werror`. It took me about a week to figure out where that particular warning went :-)

> UTScapy and pytest should not coexist

I think it would be easier to switch to pytest if it was possible to convert the tests one by one.

**Comment by polybassa** [comment]
Hello together, I would like to come back to this issue. I'm wondering, if we could start supporting pytest, in order to have to opportunity to start a transition from UTScapy to pytest.

**Comment by guedou** [comment]
I still believe that it is a good idea to move to pytest. However, I am scared by the amount of work and review time as well as the fear that we will stop the migration before it is fully completed.We could start by splitting regression.uts into smaller pieces and use this opportunity to store uts files under the same architecture as scapy/Then, we could start the migration step by step by submitting small PR.What do you think?Sent from my iPhoneOn 6 Jun 2023, at 13:04, Nils Weiss ***@***.***> wrote:﻿
I've made a few tests, apparently chatGPT is really useful to translate unit tests from UTScapy to pytest. This would allow a transition from UTScapy to pytest with a fraction of the effort, compared to doing everything manually.
